### ISSUE TITLE
UX/UI improvement when cancelling setup on Google authorisation page

### ISSUE THREAD

**Issue by wpdarren** [google/site-kit-wp#3160]
## Bug Description

This is more of a UI/UX improvement, rather than a bug.  When connecting your Google account for Site Kit, if you cancel the setup in the Google Authorisation screen ([Screenshot](https://images.zenhubusercontent.com/5f99965247ef5a1da6985d7d/89a64fa3-fb2d-4a6c-b66a-851b2fefc3de)), you are redirected to WordPress Dashboard and receive a message `The Site Kit setup was interrupted because you did not grant the necessary permissions. To fix this, redo the plugin setup.` 

![Untitled](https://user-images.githubusercontent.com/1621608/116210796-e0277480-a74b-11eb-991c-851a55946b96.jpg)

I am wondering if we should be consistent because, when you cancel the set up within the sitekit.withgoogle.com pages ([Screenshot](https://images.zenhubusercontent.com/5f99965247ef5a1da6985d7d/9df18c48-defc-498d-bfc2-42ea387e3934)) the user is redirected to the Site Kit dashboard and the Set up Site Kit screen appears. 

![image.png](https://images.zenhubusercontent.com/5f99965247ef5a1da6985d7d/329d763b-58e4-4694-94dd-0a41d949fe13)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* When something fails during the OAuth flow, the user currently gets redirected to the WP Admin dashboard (`admin_url()`). They should instead be redirected to the Site Kit dashboard (if already set up/connected, i.e. if the user can already `VIEW_DASHBOARD`) or splash screen.
* The current `googlesitekit_error_code`, its respective error message, and a potential "proxy access code" should be made available in JS. If one is set, both the dashboard and splash screen should include logic to display a JS notification on top of the screen.
    * It should have the following heading: _Error connecting Site Kit_
    * The main content should be the error message.
    * There should be a primary CTA button to "Redo the plugin setup" which takes the user to the proxy setup URL (which needs to include the proxy access code if available, see current `Authentication::get_authentication_oauth_error_notice` method).
* The PHP notice from `Authentication::get_authentication_oauth_error_notice` should be removed, since it is replaced with the JS notification.

## Implementation Brief

**Migrate `global._googlesitekitLegacyData.setup.errorMessage` to the datastore, and add the "redo" proxy setup URL to the datastore if the proxy access code is available.**

Within `includes/Core/Authentication/Authentication.php`:
* Move the code that sets `errorMessage` from the `inline_js_setup_data` function to the `inline_js_base_data` function. Rename `errorMessage` to `setupErrorMessage`.
https://github.com/google/site-kit-wp/blob/487c88b63659dd74ddfe5954eb9edbf48de2de73/includes/Core/Authentication/Authentication.php#L820-L823
* At the same place where `setupErrorMessage` is now set:
  * Retrieve the access code: `$access_code = $this->user_options->get( OAuth_Client::OPTION_PROXY_ACCESS_CODE );`
  * If the access code is set, then add the proxy setup URL to `$data`. Name the field `setupErrorRedoURL`.
    * This code snippet can be used to retrieve the URL (note, this depends on #4588 being merged):
https://github.com/google/site-kit-wp/blob/0305b070dacf5fb4091091499ad3bf8cce6c77ed/includes/Core/Authentication/Authentication.php#L1099-L1108
    * Then set `$data['setupErrorRedoURL'] = $setup_url`.
  * Delete the stored access code: `$this->user_options->delete( OAuth_Client::OPTION_PROXY_ACCESS_CODE );`.

Within `assets/js/googlesitekit/datastore/site/info.js`:

* Add `setupErrorMessage` and `setupErrorRedoURL` to the reducer (in the `case: RECEIVE_SITE_INFO` block, extract them from `payload.siteInfo` and return in the `siteInfo` state object).
* Add `setupErrorMessage` and `setupErrorRedoURL` to the `*getSiteInfo` resolver (extracting them from `global._googlesitekitBaseData`).
* Add selectors `getSetupErrorMessage` and `getSetupErrorRedoURL`, following the existing pattern using the `getSiteInfoProperty` utility function.

Within `assets/js/components/setup/SetupUsingProxy.js`:

* Replace this line which retrieves `errorMessage` with a `useSelect` which returns `select( CORE_SITE ).getSetupErrorMessage()`:
https://github.com/google/site-kit-wp/blob/487c88b63659dd74ddfe5954eb9edbf48de2de73/assets/js/components/setup/SetupUsingProxy.js#L157-L158

**Redirect to the dashboard or splash screen on OAuth error.**

Within `includes/Core/Authentication/Clients/OAuth_Client.php`:

* Create a new method, called `authorize_user_redirect_url`, that returns the dashboard URL if the user has permission, otherwise the splash URL:
```php
	private function authorize_user_redirect_url() {
		return current_user_can( Permissions::VIEW_DASHBOARD )
			? $this->context->admin_url( 'dashboard' )
			: $this->context->admin_url( 'splash' );
	}
```
* Within the `authorize_user` method, replace the calls to `admin_url()` with `authorize_user_redirect_url()`.

**Add CTA to the error notification on the splash page.**

Within `assets/js/components/setup/SetupUsingProxy.js`:

* Retrieve `setupErrorRedoURL` via a `useSelect` which returns `select( CORE_SITE ).getSetupErrorRedoURL()`.
* Add `ctaLink` and `ctaLabel` props to the `BannerNotification` that is shown for an `errorMessage`:
```js
ctaLink={setupErrorRedoURL}
ctaLabel="Redo the plugin setup"
```

**Add error notification with CTA to the dashboard (replacing `UnsatisfiedScopesAlert`).**

Within `assets/js/components/notifications/ErrorNotifications.js`:

* Retrieve `setupErrorMessage` and `setupErrorRedoURL` via their selectors as described above.
* Replace the rendered `UnsatisfiedScopesAlert` with a conditional render. If `setupErrorMessage` is set, render a `BannerNotification`, otherwise continue to render `UnsatisfiedScopesAlert`:
```js
{ setupErrorMessage ? (
	<BannerNotification
		id="setup_error"
		type="win-error"
		title={ __(
			'Error connecting Site Kit.',
			'google-site-kit'
		) }
		description={ setupErrorMessage }
		isDismissible={ false }
		ctaLink={setupErrorRedoURL}
		ctaLabel="Redo the plugin setup"
	/>
) : (
	<UnsatisfiedScopesAlert />
) }
```

**Remove PHP-side notice**

Within `includes/Core/Authentication/Authentication.php`:

* Remove the `get_authentication_oauth_error_notice` method.
* Update `authentication_admin_notices` to no longer attempt to call `get_authentication_oauth_error_notice`.

### Test Coverage

* Fix any failing tests.
* Update the tests for `assets/js/googlesitekit/datastore/site/info.js` to reflect the new properties added in this file.

### Visual Regression Changes

* 

## QA Brief

* Start the Site Kit setup, but cancel setup when on a [permissions screen](https://images.zenhubusercontent.com/5f99965247ef5a1da6985d7d/89a64fa3-fb2d-4a6c-b66a-851b2fefc3de). You should see an error message with the text from the ACs and a "Redo the plugin setup" button that resumes the plugin setup.

Try the above for a user who has previously authenticated and a user that has never authenticated for the website in question.

## Changelog entry

* Redirect back to Site Kit when an error occurs during Google sign-in instead of the WordPress dashboard.

**Comment by aaemnnosttv** [comment]
@felixarntz any reason why we can't do this?

**Comment by felixarntz** [comment]
@aaemnnosttv This is a valid improvement. It should be trivial, as it looks like we already handle any OAuth error code present and output the respective error on the splash screen. The logic to include the `errorMessage` JS field that is used for this is only run for sites using the proxy though, so maybe we should make this conditional (keeping the old state for the <1%).

Maybe this is also a good point to migrate `global._googlesitekitLegacyData.setup.errorMessage` to the datastore instead.

**Comment by eclarke1** [comment]
@asvinb is this an IB your are aiming to continue with, or could you be unassigned for someone else to pick up? Happy either way, just wanted to check.

**Comment by techanvil** [comment]
Hi @felixarntz, I've been looking into this ticket and have noticed there is some existing behaviour that we might need to account for.

I've locally modified [`OAuth_Client::authorize_user`](https://github.com/google/site-kit-wp/blob/f864bbc681ec362492621a78841b1aea4ba87854/includes/Core/Authentication/Clients/OAuth_Client.php#L368) to redirect to the Splash Page or Dashboard (if permissioned, i.e. for a module setup) when there is an OAuth error.

I've noticed that the Splash Page then shows the error message as a [`BannerNotification`](https://github.com/google/site-kit-wp/blob/develop/assets/js/components/setup/SetupUsingProxy.js#L216-L217), along with the usual CTA below **Set up Site Kit**.

![image](https://user-images.githubusercontent.com/18395600/147083099-6989dcb7-e07b-446d-a1a2-aea90383ee5b.png)

While the Dashboard displays the [`UnsatisfiedScopesAlert`](https://github.com/google/site-kit-wp/blob/f864bbc681ec362492621a78841b1aea4ba87854/assets/js/components/notifications/UnsatisfiedScopesAlert.js) error message, along with its CTA to redo the setup (the same occurs on the non-unified Dashboard).

![image](https://user-images.githubusercontent.com/18395600/147082348-1576908b-c6e6-467a-a274-d5e75b3df8ae.png)

I'm therefore wondering how this affects the requirements here. Do we still need to change the UI on the Splash Page? At present it is showing the error message along with a CTA which does seem to fit the spirit if not the letter of the requirements. The Dashboard is also showing a CTA, and an error message, albeit not the actual OAuth error message itself. However I'm thinking that an additional notification would start to look a bit cluttered. Should we keep the `UnsatisfiedScopesAlert`, or modify or replace it?

Interested to hear your thoughts.

**Comment by felixarntz** [comment]
@techanvil Great questions! I think on the splash screen the current implementation has us indeed _almost_ covered - what still needs to be added here is support to show a blue retry link within that notification _if_ a "proxy access code" is also available (potentially passed from the proxy together with the error code).

On the Site Kit dashboard, I think we should then implement the same banner notification per the ACs, displaying the message and button if the "proxy access code" is there. I think in that case (_if_ that notification is shown on the dashboard), we can have it replace the `UnsatisfiedScopesAlert`. The latter should however still show in any other case when the new notification is not present, e.g. if the user decides to not act on the new notification and then visits the dashboard again.

**Comment by aaemnnosttv** [comment]
> Does this mean we should in fact keep `Authentication::get_authentication_oauth_error_notice`, possibly adding a condition around it?

@techanvil I don't think so – the `errorMessage` JS field refers to this:
https://github.com/google/site-kit-wp/blob/6e06d608c064e0ee8313a064cc6e9b4236da6915/includes/Core/Authentication/Authentication.php#L820-L825

which is only used here at the moment (as you identified):
https://github.com/google/site-kit-wp/blob/6e06d608c064e0ee8313a064cc6e9b4236da6915/assets/js/components/setup/SetupUsingProxy.js#L157-L158

So that wouldn't affect the notice in PHP one way or the other. I think perhaps @felixarntz was referring to making it conditionally available in JS based on `usingProxy` as it is today (conditionally exposing the value). This comment was made long before the ACs were written though and isn't mentioned in the ACs so I wouldn't worry about it for now.

**Comment by techanvil** [comment]
Thanks @aaemnnosttv - I think then, we can simply remove `Authentication::get_authentication_oauth_error_notice`. I've updated the IB accordingly.

**Comment by felixarntz** [comment]
@techanvil IB ✅  Just one tiny thing: Let's rename `redoProxySetupUrl` in a way that it is clear that this belongs with `setupErrorMessage` - for example `setupErrorRedoURL` or something like that. If you agree, can you update the IB like this? Then feel free to move to _Execution Backlog_ right away.

**Comment by techanvil** [comment]
Thanks @felixarntz, that sounds sensible. I have renamed it to `setupErrorRedoURL` (and related `getSetupErrorRedoURL`) as suggested, and moved this over to _Execution Backlog_.

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@tofumatt I have a few issues, detailed below:

1.  The title `Error Connecting Site Kit` has a period at the end. 

![image](https://user-images.githubusercontent.com/73545194/170969152-4b584b8f-faf0-4d56-93fe-63409954a5f7.png)

2. When setting up Idea Hub.  Click the CTA button (or link in settings) to set the feature up, and then do not give permission (or cancel) and you are redirected back to the Site Kit Dashboard with the error message `Error Connecting Site Kit` Click on the `Redo the Plugin setup` button and you go through the authorisation process and this time give permission.  You are redirected back to the dashboard, but as you can see, the set up hasn't completed.  

The Idea Hub step where you have to agree for Idea Hub to track your usage of the feature does not appear. 

https://user-images.githubusercontent.com/73545194/170970508-a9f16504-1008-4df2-9b71-0373ebd14c4e.mp4

3. The same issue above was also occurring for Analytics but the only thing different this time is that when I was redirected back to the dashboard, the Analytics success notification appeared but you can see from the screencast below, it hasn't. 

https://user-images.githubusercontent.com/73545194/170971072-5da5e572-2f4a-48ee-816c-fbed88d121b6.mp4

Both of these modules were set up after initially connecting Site Kit with Search Console.  The issues were tested with a new site, and a Google account that has had no permissions previously.  

4. This is more of an observation from a UI/UX perspective. As you can see from the two screenshots below, there are other CTA buttons underneath the error.  Should the error just be loaded with no widgets or any splash page underneath?  I just wonder if this could cause confusion with users. 

<details>
 <summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/73545194/170971900-ee1f53c8-ac35-4d83-a959-48ef2420746d.png)

![image](https://user-images.githubusercontent.com/73545194/170971932-ab77de77-2373-4a2d-8b23-1907b44dccf8.png)

</details>

**Comment by tofumatt** [comment]
Good catch on the header, I've filed a follow-up PR for that #5303. 👍🏻 

It seems like that reauth bug for specific modules is long-standing and not actually caused by this issue. Here's me running through the same flow you mentioned in 1.75:

https://user-images.githubusercontent.com/90871/171068720-66ee398d-d727-4ff4-86c1-fcb6b298e97c.mp4

No doubt a problem, but I think one worth filing a separate issue for because anytime a module tries to get you to reauth it seems we get the OAuth info but direct the user back to the dashboard rather than the module setup page.

I think it's worth fixing but out-of-scope for this issue.

@wpdarren Can you file a new issue with the steps-to-reproduce the module "redo setup" bugs you mention in points 2/3?

For point 4—I think this is intentional, because we want a related CTA button to the error. It's okay both buttons allow the user to redo the setup, but I don't think hiding the content is the right choice here.
